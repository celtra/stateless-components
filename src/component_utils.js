const flatVariations = (variations) => {
    let flat = [{}]

    for (const key of Object.keys(variations).sort((a, b) => a.localeCompare(b))) {
        let newFlat = []
        for (const value of variations[key]) {
            newFlat = newFlat.concat(flat.map(item => {
                return { ...item, [key]: value }
            }))
        }
        flat = newFlat
    }

    return flat
}

const getHash = (usecase) => {
    const s = Object.keys(usecase)
        .sort((a, b) => a.localeCompare(b))
        .filter(key => typeof usecase[key] !== 'function')
        .map(key => {
            const value = typeof usecase[key] === 'object' ? JSON.stringify(usecase[key]).trim('"') : usecase[key]
            return `${key}=${value}`
        })
        .join('&')

    var a = 1, c = 0, h, o
    if (s) {
        a = 0
        for (h = s.length - 1; h >= 0; h--) {
            o = s.charCodeAt(h)
            a = (a<<6&268435455) + o + (o<<14)
            c = a & 266338304
            a = c!==0?a^c>>21:a
        }
    }
    return String(a)
}

export function getFlatUsecases (component, ignoreVariations = []) {
    const usecases = []
    if (component.usecases) {
        const variations = component.variations && component.variations || {}
        const filterVariations = {}
        for (const k in variations) {
            if (!ignoreVariations.includes(k)) {
                filterVariations[k] = variations[k]
            }
        }

        for (const variation of flatVariations(filterVariations)) {
            for (const usecase of component.usecases) {
                const usecaseData = { ...variation, ...usecase }
                usecases.push({ data: usecaseData, uniqueID: getHash(usecaseData) })
            }
        }
    }
    return usecases
}

export function getProps (component) {
    const componentProps = component.props
    const usecases = getFlatUsecases(component)
    const defaultProps = usecases.length > 0 ? usecases[0].data : {}

    let allProps = {}
    for (const key in componentProps) {
        allProps[key] = true
    }
    for (const key in defaultProps) {
        allProps[key] = true
    }
    allProps = Object.keys(allProps)

    return allProps.map(propName => {
        return {
            name: propName,
            type: componentProps.hasOwnProperty(propName) ? componentProps[propName].type : typeof defaultProps[propName],
            default: defaultProps.hasOwnProperty(propName) ? defaultProps[propName] : componentProps[propName].default,
            availableValues: component.variations && component.variations[propName] || null,
        }
    })
}
