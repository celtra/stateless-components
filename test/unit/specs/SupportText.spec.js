import SupportText from '@/stateless/support_text.vue'
import Vue from 'vue'
import { getFlatUsecases } from '@/component_utils'

const Constructor = Vue.extend(SupportText)
const vm = new Constructor({
    propsData: getFlatUsecases(SupportText)[0].data,
}).$mount()

describe('SupportText', () => {
    it('should match the snapshot', () => {
        expect(vm.$el).toMatchSnapshot()
    })
})
