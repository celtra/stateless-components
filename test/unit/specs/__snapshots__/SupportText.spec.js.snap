// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SupportText should match the snapshot 1`] = `
<div
  class="support-text"
>
  <div
    class="support-text__main support-text__main--left"
  >
    <!---->
     
    <div
      class="support-text__icon"
    >
      <svg
        viewBox="0 0 17 17"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
        >
          <circle
            cx="8.5"
            cy="8.5"
            fill="#D8D8D8"
            r="8.5"
          />
           
          <path
            d="M6 6.505h1.746c.016-.52.37-.864.886-.864.505 0 .86.3.86.737 0 .473-.19.7-.92 1.128-.746.43-1.014.913-.912 1.832l.016.215h1.676V9.3c0-.505.204-.746.94-1.17.79-.457 1.16-1.005 1.16-1.8 0-1.29-1.064-2.148-2.697-2.148-1.718 0-2.74.918-2.755 2.325zm2.562 5.683c.693 0 1.096-.36 1.096-.978 0-.623-.403-.982-1.096-.982-.693 0-1.1.36-1.1.982 0 .618.407.978 1.1.978z"
            fill="#1F1F2C"
          />
        </g>
      </svg>
    </div>
     
    <!---->
  </div>
</div>
`;
